keiko-sql/src/main/kotlin/com/netflix/spinnaker/q/sql/SqlQueue.kt

```diff
@@ -402,24 +403,46 @@ class SqlQueue(
       jooq.transaction { config ->
         val txn = DSL.using(config)
 
-        txn.insertInto(messagesTable)
-          .set(idField, ulid.toString())
-          .set(fingerprintField, fingerprint)
-          .set(bodyField, mapper.writeValueAsString(message))
-          .set(updatedAtField, clock.millis())
-          .onDuplicateKeyUpdate()
-          .set(idField, MySQLDSL.values(idField) as Any)
-          .set(bodyField, MySQLDSL.values(bodyField) as Any)
-          .execute()
+        val existingFingerprint = txn.select(field(idField))
+          .from(unackedTable)
+          .where(fingerprintField.eq(fingerprint))
+          .fetchOne(0, String::class.java)
```

Comment:
Since this is in a transaction, keep an eye on whether this increases the frequency of deadlocks. Its probably fine but I previously had to remove or simplify transactions to avoid lock issues under load. Another option would be to do something like this outside of a txn:
```
insert into keiko_v1_queue_default (id, fingerprint, delivery, locked)
	select id, fingerprint, delivery, locked
		from (select "abc" as id, "6b38d346dcd4509083ebe9954b8c0030" as fingerprint, 0 as delivery, "0" as locked) x
		where not exists (select 1 from keiko_v1_unacked_default where keiko_v1_unacked_default.fingerprint = "6b38d346dcd4509083ebe9954b8Zc0030")
```
(taken from https://stackoverflow.com/questions/43804631/mysql-conditional-insert-to-one-table-if-a-certain-row-does-not-exist-on-another and tested against a local orca/keiko db)

With that, you'd look at the number of rows changed that jooq returns to branch from there. I.e. 0 == on the unacked queue, 1 == not on unacked and inserted to queue.

Reply:
Actually, since existingFingerprint isn't set via a locking read and orca doesn't use repeatable reads, I think there's still a race condition where the message is already queued and moved to unacked by another instance in-between the `existingFingerprint` and the later write. I'm pretty sure making that a select a "for update" inside this txn will cause performance issues from earlier versions of keiko. The hack to implement "insert .. where" looks like it'd be safer.

Reply:
Thanks for the tips!